Bluetooth: Mesh:  Add disabling/enabling GATT services to suspend/resume

HaavardRei opened the pull request:
PR contains four commits:

- Extracts some GATT related logic from `test_advertiser.c` bsim test into a common file. 
- Refactors `test_suspend.c` by reusing a RX test config. 
- Disables/enables PB-GATT and GATT proxy advs when suspending/resuming Mesh. 
- Adds tests to `test_suspend.c` to check that GATT advs stop when suspending Mesh, and that they start again when resuming Mesh.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -290,6 +408,131 @@ static void test_rx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(6500));
+
+	/* Send a mesh message to notify tester that device is suspended */
+	struct bt_mesh_adv *adv = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_ADV_TAG_LOCAL,
+			BT_MESH_TRANSMIT(0, 20), K_NO_WAIT);
+	net_buf_simple_add_mem(&adv->b, "1", sizeof("1"));
+	bt_mesh_adv_send(adv, NULL, NULL);
+	k_sleep(K_MSEC(150));
```
Wrap into a function?

Andrewpini replied:
This is purely used as a sync mechanism right? From what I can see we do not have anything like this in the common modules for bsim tests. IMO if we are adding this sort of sync mechanism, it should be added to the common library so we don't end up re-implementing it for other tests. I have considered something like this before myself.

HaavardRei replied:
Yes, only used as a sync mechanism. 
I will look into adding it to the common lib. Should we have a fixed message size (i. e. 8/16-bit) and retransmit count, or let these be specified by the user?

HaavardRei replied:
I added two functions to `mesh_test.c/h` for the message passing/sync. May be hard to look at the diff since I also did a rebase, but the changes are in `mesh_test`, `gatt_common` and the test scripts where I use the updated functions.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,6 +171,135 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
+	k_sem_give(&publish_sem);
+}
+
+static void gatt_rx_xmit_adv(void)
```
```suggestion
static void suspend_state_change_wait(void)
```
or smth similar? `gatt_rx_xmit_adv` is vague.

HaavardRei replied:
Changed!

Andrewpini commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,6 +171,135 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
+{
+	static int cnt;
+
+	LOG_INF("rx: cnt(%d)", cnt);
```
In general I can appreciate adding more log to the Bsim tests 👍 But I think most of the `LOG_INF` can be converted to `LOG_DBG`. For most iterations of this test most of this log is not relevant. If anybody needs to take a closer look they can modify `LOG_MODULE_REGISTER(test_suspend, LOG_LEVEL_INF);` to enable this logging.

Andrewpini commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -290,6 +408,131 @@ static void test_rx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(500));
+	gatt_rx_beacons();
+
+	/* Wait for dut to resume Mesh and notify tester, then scan for gatt proxy beacons */
```
```suggestion
	/* Wait for DUT to resume Mesh and notify Tester, then scan for gatt proxy beacons */
```

Since we are talking of DUT and tester, I would prefer if you updated the test declaration. I know `tx` and `rx` are widely used, but in this instance I belive that `dut` and `tester` provides better insight for the reader. Also, I'd prefer something a bit more descriptive in the test string. Maybe something like this?:
```
static const struct bst_test_instance test_suspend[] = {
	TEST_CASE(dut, suspend_resume, "Suspend and resume Mesh while advertising periodic pub"),
	TEST_CASE(dut, suspend_disable_resume,
		  "Suspend and resume Mesh (and disable/enable BT) while advertising periodic pub"),
	TEST_CASE(dut, gatt_suspend_resume, "Suspend and resume Mesh while advertising GATT proxy"),
	TEST_CASE(dut, gatt_suspend_disable_resume,
		  "Suspend and resume Mesh (and disable/enable BT) while advertising GATT proxy"),

	TEST_CASE(tester, pub, "Scan and verify behavior of periodic pub adv"),
	TEST_CASE(tester, gatt, "Scan and verify behavior of GATT proxy adv"),

	BSTEST_END_MARKER};
```

alxelax commented on tests/bsim/bluetooth/mesh/src/gatt_common.h:
```diff
@@ -0,0 +1,28 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
```
`mesh_test.h` has already included kernel header.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,6 +171,135 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
+
+	char *data = net_buf_simple_pull_mem(buf, sizeof("0"));
+
+	if (strncmp(data, "0", strlen("0")) == 0) {
```
It extracts 2 bytes from payload but compares 1. Is this mistake?
Just FYI:
```
#include <stdio.h>
#include <string.h>

int main()
{
    printf("strlen: %li \n", strlen("0"));
    printf("sizeof: %li \n", sizeof("0"));

    return 0;
}
```
```
strlen: 1 
sizeof: 2 
```

alxelax replied:
Does it have sense to use string related functions to compare 1-2 bytes variables?

HaavardRei replied:
Changed to status message to an enum and adapted size.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,6 +171,135 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
+	ASSERT_FALSE_MSG(err && err != -EALREADY, "Stopping scan failed (err %d)\n", err);
+}
+
+static void gatt_xmit_scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type,
```
Is this really gatt? `BT_GAP_ADV_TYPE_ADV_NONCONN_IND` is type of frames over adv bearer.
It just parses ad structure. There is no any id of gatt services.

HaavardRei replied:
I mainly used the gatt prefix to show that the cb relates to the gatt tests of suspend/resume, but I will rename it.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -290,6 +408,131 @@ static void test_rx_suspend_disable_resume(void)
+	/* Send a mesh message to notify tester that device is suspended */
+	struct bt_mesh_adv *adv = bt_mesh_adv_create(BT_MESH_ADV_DATA, BT_MESH_ADV_TAG_LOCAL,
+			BT_MESH_TRANSMIT(0, 20), K_NO_WAIT);
+	net_buf_simple_add_mem(&adv->b, "1", sizeof("1"));
```
It is better to use meaningful messages rather than 0 or 1. Smth. like `suspend` and `resume` to have self-descriptive source code that doesn't require comments.

HaavardRei replied:
Changed to enum.

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -290,6 +408,131 @@ static void test_rx_suspend_disable_resume(void)
+	ASSERT_OK_MSG(bt_enable(NULL), "Bluetooth init failed");
+	LOG_INF("Bluetooth initialized");
+
+	/* Scan pb gatt beacons. */
```
`bt_mesh_suspend` will fail if the device is unprovisioned:
https://github.com/zephyrproject-rtos/zephyr/blob/ecce23532270a0b5bb6e64f97503b1b3cf00b26a/subsys/bluetooth/mesh/main.c#L438-L440

We might not want this behavior? If so, I can create a task for it.

HaavardRei commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -508,6 +516,14 @@ int bt_mesh_resume(void)
 	}
 
+	if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)) {
+		(void)bt_mesh_proxy_gatt_enable();
```
See above comment

HaavardRei commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -462,6 +462,14 @@ int bt_mesh_suspend(void)
 	bt_mesh_access_suspend();
 
+	if (IS_ENABLED(CONFIG_BT_MESH_PB_GATT)) {
+		(void)bt_mesh_pb_gatt_srv_disable();
```
When resuming mesh, `bt_mesh_proxy_gatt_enable()` returns `1`. This is due to the return value of [k_work_schedule()](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#c.k_work_schedule)  being returned.
This is not really an error and should be ignored?

alxelax replied:
Suspending\resuming functionality shouldn't know implementation details of other modules. There are API functions. These should be sufficient. If API function behaves incorrectly, then it should be fixed inside the module. There shouldn't be workarounds here for adjusting internal implementation.
Also, suspending\resuming functionality shouldn't predict reasons of returned values\errors. Today it returns errors because device is unprovisioned, tomorrow somebody adds new functionality and new statuses.

HaavardRei replied:
It now returns error codes. Added a separate commit after offline discussion fixing the return value of bt_mesh_proxy_gatt_enable `bt_mesh_proxy_gatt_enable`

HaavardRei committed:
bsim: Bluetooth: Mesh: extract common adv logic

Extracts logic related to waiting for packets from
`test_advertiser.c` and `test_beacon.c` to use in a sync mechanism.
Introduces a sync mechanism in bsim with the functions
`bt_mesh_test_send_over_adv` and `bt_mesh_test_wait_for_packet`

Extracts gatt-related logic from `test_advertiser.c` into a separate
file so it can be re-used.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
bsim: Bluetooth: Mesh: refactor suspend test

Renames tx to DUT and rx to Tester to clarify roles in test.

Re-uses the same body for suspend/resume and suspend/disable/resume
for DUT with a parameter to toggle disabling BT.

Re-uses the same Tester config in both existing test-cases as the
previous configs were duplicates.

Updates the dut suspension status in the message handler.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
bsim: Bluetooth: Mesh: Add gatt suspension test

Adds two test-cases to `test_suspend.c`, both checking that gatt
advertisement is stopped when suspending Mesh, and that is started again
when Mesh is resumed.

The first test-case involves suspending and resuming Mesh, and
the second test-case involves suspending Mesh and disabling Bluetooth,
then re-enabling Bluetooth and resuming Mesh.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Mesh: suspend/resume gatt advs

Disables pb gatt- and gatt proxy advs when suspending Mesh, and
enables them again when resuming Mesh.

Adds `bt_mesh_adv_gatt_send` to `bt_mesh_resume` to make sure that
GATT advs start after resumption.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

alxelax commented on tests/bsim/bluetooth/mesh/src/mesh_test.h:
```diff
@@ -23,6 +23,8 @@
 
 #include <zephyr/bluetooth/bluetooth.h>
 #include <zephyr/bluetooth/mesh.h>
+#include <zephyr/bluetooth/hci.h>
```
Why do you need hci header?

HaavardRei replied:
Used in `bt_mesh_test_wait_for_adv` in `mesh_test.c`.

```
...
  struct bt_le_scan_param scan_param = {
    .type       = BT_HCI_LE_SCAN_PASSIVE,
    .options    = BT_LE_SCAN_OPT_NONE,
    .interval   = BT_MESH_ADV_SCAN_UNIT(1000),
    .window     = BT_MESH_ADV_SCAN_UNIT(1000)
  };
```

alxelax replied:
If this is required only for provided above code why it is here instead of c file?

alxelax commented on tests/bsim/bluetooth/mesh/src/mesh_test.c:
```diff
@@ -546,6 +546,46 @@ uint16_t bt_mesh_test_own_addr_get(uint16_t start_addr)
+	bt_mesh_adv_send(adv, NULL, NULL);
+}
+
+int bt_mesh_test_wait_for_adv(struct k_sem *observer_sem, bt_le_scan_cb_t scan_cb)
```
It waits for any frame, not only adv. Scanner with these settings can receive gatt frames too.

HaavardRei replied:
Will look into something similar to beacon test.

alxelax replied:
As I wrote above, the function name is not correct. It receives not only over advertising but over gatt too.
It should be something like: `bt_mesh_test_wait_for_packet` or `bt_mesh_test_wait_for_frame`.

HaavardRei replied:
Changed to `bt_mesh_test_wait_for_packet`

alxelax commented on tests/bsim/bluetooth/mesh/src/test_advertiser.c:
```diff
@@ -282,26 +237,6 @@ static void xmit_scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type
 	}
 }
 
-static void rx_xmit_adv(void)
```
Why did you move `rx_xmit_adv` in common place but left `rx_gatt_beacons` and `receive_order`?
It should be unified in general manner or shouldn't be touched at all. Right now, the same behavior is located in two different files: in general file and in test scenario implementation.

It is better to concentrate on your target rather to start refactoring of another tests in the same PR that devoted to suspend resume functionality.

HaavardRei replied:
The implementation of `receive_order` is tailored to the adv test with number of expected advs, but I agree that `rx_gatt_beacons` could be refactored to something like "`bt_mesh_test_wait_for_adv(&observer_sem, gatt_scan_cb)`". Any thoughts on this?

The thought was to extract the functionality that would be duplicate between the two tests, but it seems that there was more duplicate functionality than expected.

HaavardRei replied:
I agree, I will address the other comments before looking into adding a separate commit with more refactoring :)

HaavardRei replied:
I'm now using `bt_mesh_test_wait_for_adv` in all places we have discussed, and generalized it to having variable waiting time.

alxelax replied:
Yep, currently refactoring of the scanner related code looks much better.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -62,7 +71,22 @@ static int msg_handler(const struct bt_mesh_model *model, struct bt_mesh_msg_ctx
-	ASSERT_FALSE_MSG(suspended, "Received publication while Mesh is suspended.");
+	if (dut_status == DUT_SUSPENDED) {
+		if (num == (NUM_PUB + 1)) {
+			/* A publication may be sent just before DUT is suspended. Do nothing. */
```
Strange alignment

HaavardRei replied:
Strange how? It is an empty if statement?

alxelax replied:
Why do you need empty `if` statement?

HaavardRei replied:
Just one possible solution to make tester ignore a publication that might be sent before the DUT is fully suspended.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -62,7 +71,22 @@ static int msg_handler(const struct bt_mesh_model *model, struct bt_mesh_msg_ctx
+	if (dut_status == DUT_SUSPENDED) {
+		if (num == (NUM_PUB + 1)) {
+			/* A publication may be sent just before DUT is suspended. Do nothing. */
+		} else if (SUSPEND_DURATION*1000 <= k_uptime_delta(&uptime)) {
```
```suggestion
		} else if (SUSPEND_DURATION * 1000ll <= k_uptime_delta(&uptime)) {
```

HaavardRei replied:
Changed!

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -62,7 +71,22 @@ static int msg_handler(const struct bt_mesh_model *model, struct bt_mesh_msg_ctx
+	if (num == NUM_PUB) {
+		dut_status = DUT_SUSPENDED;
+		LOG_DBG("Expecting number of pubs received. Setting status to %d.", dut_status);
+		uptime = k_uptime_get();
```
uptime has been already updated before.

HaavardRei replied:
This is the only call to `k_uptime_get`, and the uptime set here is used to check that the suspend duration has passed when the DUT starts sending messages again.

HaavardRei replied:
This is the first time uptime is set during the test. The uptime is updated by `k_uptime_delta`, but at this point the status will either get updated to `DUT_RUNNING` (suspend duration passed) or the test will fail due to receiving a message during suspension.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,79 +189,127 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
 }
 
-static void test_tx_suspend_disable_resume(void)
+static void dut_pub_body(bool disable_bt)
```
```suggestion
static void dut_pub_common(bool disable_bt)
```

HaavardRei replied:
Changed to your suggestion, same for below.

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -270,15 +344,76 @@ static void test_tx_suspend_disable_resume(void)
 	ASSERT_OK(bt_mesh_suspend());
+}
+
+static void dut_gatt_body(bool disable_bt)
```
```suggestion
static void dut_gatt_common(bool disable_bt)
```

Andrewpini commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -270,15 +344,76 @@ static void test_tx_suspend_disable_resume(void)
+
+	/* Send a mesh message to notify Tester that DUT is about to be suspended. */
+	dut_status = DUT_SUSPENDED;
+	bt_mesh_test_send_over_adv(&dut_status, sizeof(enum dut_mesh_status));
+	k_sleep(K_MSEC(150));
```
I see that `k_sleep(K_MSEC(150))` is placed behind all instances of `bt_mesh_test_send_over_adv`. Any reason to not put it inside the function it self? Or, if this sleep is to ensure that the message has been send by the adv layer, this could be handled by using a semaphore and the end_cb of  `struct bt_mesh_send_cb` inside `mesh_test.c`. I think that would have been a little bit cleaner if this is the intention.

HaavardRei replied:
It's to ensure that the adv/message is sent before suspending DUT. It's only really relevant in the cases where we suspend just after sending, so I suggest we keep it as is. 

I can add a comment to this case to make it more clear?

Andrewpini replied:
Ok, sure. I would argue that for a sync mechanism this would be relevant for a lot of usecases, but it is not critical.

Andrewpini commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -165,79 +189,127 @@ struct bt_mesh_cfg_cli_mod_pub pub_params = {
+	if (err && (dut_status == DUT_RUNNING)) {
+		FAIL("Tester timed out");
+	} else if ((err == -EAGAIN) && (dut_status == DUT_SUSPENDED)) {
+		LOG_WRN("Tester timed out (DUT is suspended)");
```
I see this trigger on all GATT tests. Does it then qualify as a warning from the test POW if it is default behavior?

HaavardRei replied:
Yeah no it doesn't need a warning. I moved the `EAGAIN` check to `test_tester_gatt` because of the generalized sync/message passing method. 

However... the generalized `bt_mesh_test_wait_for_adv` function will give a warning if waiting on the semaphore times out. This is because the function is used in multiple places (adv and beacon tests).

Andrewpini commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -290,18 +425,68 @@ static void test_rx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(1000));
+
+	/* Scan gatt proxy beacons. */
+	gatt_param.transmits = 5000 / 1000;
```
I would add/copy the `/* (total transmit duration) / (transmit interval) */` line from `test_advertiser.c` as well. It makes it easier to understand why it is written in such a way😺

HaavardRei replied:
Donezo!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/mesh_test.c:
```diff
@@ -546,6 +546,46 @@ uint16_t bt_mesh_test_own_addr_get(uint16_t start_addr)
+	bt_mesh_adv_send(adv, NULL, NULL);
+}
+
+int bt_mesh_test_wait_for_adv(struct k_sem *observer_sem, bt_le_scan_cb_t scan_cb)
```
Agreed! Changed now.

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -270,15 +344,76 @@ static void test_tx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(1800));
+
+	LOG_INF("Provisioning DUT");
+	bt_mesh_provision(test_net_key, 0, 0, 0, dut_cfg.addr, dut_cfg.dev_key);
```
Took a lap around the code and put in ASSERTs :)

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -270,15 +344,76 @@ static void test_tx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(150));
+
+	/* Let the Tester observe that proxy advertisement resumes */
+	k_sleep(K_MSEC(6000));
```
No good reason, I removed the first sleep.

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -270,15 +344,76 @@ static void test_tx_suspend_disable_resume(void)
+	k_sleep(K_MSEC(150));
+
+	ASSERT_OK(bt_mesh_suspend());
+	LOG_INF("Mesh suspended.");
```
Changed all LOG_INFs to either LOG_DBG or ASSERT_..._MSG

alxelax commented on tests/bsim/bluetooth/mesh/src/mesh_test.c:
```diff
@@ -546,6 +546,54 @@ uint16_t bt_mesh_test_own_addr_get(uint16_t start_addr)
+	err = k_sem_take(observer_sem, K_SECONDS(wait));
+	if (err == -EAGAIN) {
+		LOG_WRN("Taking sem timed out (err %d)", err);
+		err = bt_le_scan_stop();
```
Hm... could you just create one more local variable like `returned_value` and assign here to `-EAGAIN`? Or something similar.
Seems like unnecessary duplication for scanner stopping.

alxelax commented on tests/bsim/bluetooth/mesh/src/mesh_test.h:
```diff
@@ -202,6 +204,12 @@ void bt_mesh_test_ra_cb_setup(void (*cb)(uint8_t *, size_t));
 uint16_t bt_mesh_test_own_addr_get(uint16_t start_addr);
 
+void bt_mesh_test_send_over_adv(void *data, size_t len);
+/* Wait for an adv sent by another device (i. e. using `bt_mesh_test_send_over_adv`).
```
Description is not relevant. It is not for receiving advertising only but for receiving frames over gatt too.

HaavardRei replied:
Updated the description and naming. Better now?

alxelax commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -413,17 +399,7 @@ static bool wait_for_beacon(bt_le_scan_cb_t scan_cb, uint16_t wait,
-	if (err && err != -EALREADY) {
-		FAIL("stopping scan failed (err %d)", err);
-	}
+	bool received = bt_mesh_test_wait_for_adv(scan_cb, &observer_sem, wait) ? false : true;
```
```suggestion
	bool received = !bt_mesh_test_wait_for_adv(scan_cb, &observer_sem, wait);
```
:)

alxelax reviewed (changes requested):
Looks much better, but still requires some improvements.

HaavardRei committed:
Bluetooth: Mesh: fix proxy srv return value

Previously, `bt_mesh_proxy_gatt_enable` returned the return value from
`k_work_schedule`, which could be a positive (non-error) message.
Now, it only returns negative error codes (else 0).

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

carlescufi commented:
@PavelVPV PTAL
